File: client/ChatClientCore.cs
using System.Net.Sockets;
using Shared;

namespace ChatClient;

/// <summary>서버와의 채팅 TCP 연결 하나를 관리한다. 연결 유지 + 백그라운드 수신 루프.</summary>
public class ChatClientCore
{
    private TcpClient? _tcp;
    private StreamWriter? _writer;
    private StreamReader? _reader;
    private byte[] _key = Array.Empty<byte>();
    // ponytail: 재연결 직후 옛 수신 루프가 뒤늦게 종료되며 새 연결을 끊긴 걸로 오인하지 않도록
    // 연결마다 세대 번호를 매겨 자기 세대가 최신일 때만 Disconnected를 울린다.
    private int _generation;

    public event Action<ChatEnvelope>? MessageReceived;
    public event Action<List<ChatEnvelope>>? HistoryReceived;
    public event Action? Disconnected;

    public bool IsConnected => _tcp?.Connected == true;

    /// <summary>key는 미리 파일 포트의 GETKEY로 받아온 서버 세션 암호화 키.</summary>
    public async Task ConnectAsync(string host, int port, string name, byte[] key)
    {
        var myGeneration = ++_generation;
        _key = key;
        _tcp = new TcpClient();
        await _tcp.ConnectAsync(host, port);
        var stream = _tcp.GetStream();
        _reader = new StreamReader(stream, LineProtocol.NoBom);
        _writer = new StreamWriter(stream, LineProtocol.NoBom) { AutoFlush = false };

        await LineProtocol.SendAsync(_writer, new ChatEnvelope { Type = MsgType.Join, From = name }, _key);

        _ = ReceiveLoopAsync(myGeneration);
    }

    private async Task ReceiveLoopAsync(int myGeneration)
    {
        try
        {
            while (true)
            {
                var env = await LineProtocol.ReceiveAsync(_reader!, _key);
                if (env == null) break;

                if (env.Type == MsgType.History && env.History != null)
                    HistoryReceived?.Invoke(env.History);
                else
                    MessageReceived?.Invoke(env);
            }
        }
        catch
        {
            // 연결 끊김
        }
        if (myGeneration == _generation) Disconnected?.Invoke();
    }

    /// <summary>사용자 요청으로 연결을 끊는다. 재연결은 ConnectAsync를 다시 호출하면 된다.</summary>
    public void Disconnect()
    {
        _generation++; // 옛 수신 루프가 이 세대를 더 이상 최신으로 보지 않게 함
        try { _tcp?.Close(); } catch { /* 무시 */ }
        _tcp = null;
        _writer = null;
        _reader = null;
    }

    public async Task SendAsync(string text, ChatTarget target)
    {
        if (_writer == null) return;
        await LineProtocol.SendAsync(_writer, new ChatEnvelope { Type = MsgType.Chat, Target = target, Text = text }, _key);
    }

    public async Task SendImageAsync(byte[] pngBytes, ChatTarget target)
    {
        if (_writer == null) return;
        await LineProtocol.SendAsync(_writer, new ChatEnvelope { Type = MsgType.Chat, Target = target, Image = pngBytes }, _key);
    }
}

File: mac/Sources/MacChatClient/ChatClientCore.swift
import Foundation
import CryptoKit

/// client/ChatClientCore.cs와 대응. 채팅 TCP 연결 하나 + 백그라운드 수신 루프.
/// 콜백은 모두 메인 스레드에서 호출된다.
final class ChatClientCore {
    private var stream: TCPStream?
    private var key = SymmetricKey(data: Data(repeating: 0, count: CryptoUtil.keySizeBytes))

    var onMessage: ((ChatEnvelope) -> Void)?
    var onHistory: (([ChatEnvelope]) -> Void)?
    var onDisconnected: (() -> Void)?

    /// key는 미리 파일 포트의 GETKEY로 받아온 서버 세션 암호화 키.
    func connect(host: String, chatPort: Int, name: String, key: Data) throws {
        self.key = SymmetricKey(data: key)
        let s = try TCPStream(host: host, port: chatPort)
        stream = s
        try LineProtocol.send(s, env: ChatEnvelope(type: .join, from: name), key: self.key)
        Thread.detachNewThread { [weak self] in self?.receiveLoop() }
    }

    private func receiveLoop() {
        guard let s = stream else { return }
        do {
            while true {
                guard let env = try LineProtocol.receive(s, key: key) else { break }
                if env.type == .history, let hist = env.history {
                    let cb = onHistory
                    DispatchQueue.main.async { cb?(hist) }
                } else {
                    let cb = onMessage
                    DispatchQueue.main.async { cb?(env) }
                }
            }
        } catch {
            // 연결 끊김
        }
        let cb = onDisconnected
        DispatchQueue.main.async { cb?() }
    }

    func send(text: String, target: ChatTarget) {
        guard let s = stream else { return }
        try? LineProtocol.send(s, env: ChatEnvelope(type: .chat, target: target, text: text), key: key)
    }

    func sendImage(png: Data, target: ChatTarget) {
        guard let s = stream else { return }
        try? LineProtocol.send(s, env: ChatEnvelope(type: .chat, target: target, image: png), key: key)
    }

    /// 사용자가 직접 연결을 끊는다. 소켓을 닫으면 수신 루프의 블로킹 read가 풀려나오면서
    /// 자연스럽게 onDisconnected가 호출된다.
    func disconnect() {
        stream?.close()
        stream = nil
    }
}
